pennylane/_queuing.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`QueuingContext` abstract base class.
+"""
+import abc
+from collections import OrderedDict
+
+import pennylane as qml
+
+
+class QueuingContext(abc.ABC):
+    """Abstract base class for classes that exposes a queue for objects.
+
+    In PennyLane, the construction of quantum gates is separated from the specific
+    quantum node (:class:`~.BaseQNode`) that they belong to. However, including
+    logic for this when creating an instance of :class:`~.Operator` does not align
+    with the current architecture. Therefore, there is a need to use a high level
+    object that holds information about the relationship between quantum gates and
+    a quantum node.
+
+    The :class:`~.QueuingContext` class realizes this by providing access to
+    the current QNode.  Furthermore, it provides the flexibility to have
+    multiple objects record the creation of quantum gates.
+
+    The ``QueuingContext`` class both acts as the abstract base class for all
+    classes that expose a queue for Operations (so-called contexts), as well as the
+    interface to said queues. The active contexts contain maximally one QNode and
+    an arbitrary number of other contexts like the :class:`~.OperationRecorder`.
```

Comment:
I know this is not a local modification, but I would take the opportunity to clean up the docstring, list the methods that must be overwritten, and show a quick example.

Reply:
Yes, I've left this TODo as a reminder to update the docstring. But as mentioned above, this is just a first PR which will modify these files, so I'd prefer to wait until later (in particular, when we implement nesting, this docstring will need to be updated)